v2-roadmap.md

```diff
@@ -0,0 +1,201 @@
+like `pydantic`.
+
+The separation of responsibilities in the package is not clear. We have multiple JSON
+implementations, while the format specification is the same for all bindings in the
+[`CloudEvent` spec](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/formats/json-format.md):
+
+* Pydantic implementation is calling the HTTP implementation for this, adding unnecessary overhead
+* the Kafka binding uses its own implementation, creating code duplication and unnecessary maintenance efforts
+
+We should have a single `CloudEvent` class used for all bindings/formats since the event
+is the Python representation of the event data, which is the same no matter what binding
+is used. (we potentially could receive a HTTP event and forward it as it is using Kafka).
+
+The specific bindings and formats implementations should be separate from the `CloudEvent`
+class.
+
+This proposal contains some base rules for the package architecture and how we could implement
+the `CloudEvent` spec in this package.
+
+## Requirements
+
+In order to be able to transform a python CloudEvent to/from a specific
+protocol binding, this SDK has the following responsibilities:
+
+* Identifying what attributes have to be marshalled/unmarshalled (identifying the cloudevents extensions)
+* Validate required fields and generation of defaults
+* Marshalling/unmarshalling the cloudevents core and extensions fields
+* Marshalling/unmarshalling the `data` field
+* Encode/decode the marshalled fields to/from the specific binding format
+
+## Modules
+
+We should separate the implementation of the requirements so that they can be tested independently and
+each component be used for composition.
+
+Ideally we should have these modules (naming TBC):
+
+* `cloudevents.event` - This would contain the `CloudEvent` base class. This class is the same for all bindings/formats.
+* `cloudevents.formats` - This would contain the structured formats implementations (JSON, AVRO, Protobuf)
+* `cloudevents.bindings` - This would contain the specific binding implementations (kafka, http)
+
+This matches the naming and concepts we have in the `CloudEvent` spec
+
+## Matching requirements and modules
+
+### The `cloudevents.event` module and `CloudEvent` class
+
+The `CloudEvent` class would satisfy the following requirements:
+
+* Validate required fields and generation of default values
+* Marshalling/unmarshalling the `data` field
+* Identifying what attributes have to be marshalled/unmarshalled
+
+The `data` field depends on the specific event and the `contentdatatype` attribute and
+its marshalling/unmarshalling logic should be the same for any binding/format. It seems
+a `CloudEvent` class responsibility to take care of this. (We should implement this logic
+in overridable methods)
+
+Ideally the end user will be able to extend the `CloudEvent` class for their necessity. Ie:
+
+* adding new extensions fields
```

Comment:
> There may be special extension-specific mixins that provide additional behavior

Yes,  I used the word "extend" in a generic meaning. It could be also a mixin with some custom implementation.

The idea is that it should be possible (and easy) to create a `class MyBaseEvent(MyMixin, CloudEvent)` at application level to extend the sdk functionalities.